---
title: "RNAseq Pipeline Comparison"
author: "JD"
date: "2023-06-26"
output: html_document
---

## Introduction

This RMD contains scripts and figures to compare the van Heesch RNA-seq pipeline,
both for CentOS7 module style and for Rocky8 with apptainer containers

```{r libraries}
library(rtracklayer)
library(magrittr)
library(ggplot2)
```

```{r parameters}
wd="/hpc/pmc_vanheesch/projects/Jip/pilots/20230606_JD_containers/01_rnaseq"

module_loc = paste(wd,"analysis","base",sep="/")
container_loc = paste(wd,"analysis","container",sep="/")

savedir = paste(wd,"results","comparison_analysis",sep="/")
```

## Trimming

```{r cutadapt container}

trim_stats <- read.delim(paste(container_loc,
                               "multiqc",
                               "rnaseq_container_multiqc_data",
                               "multiqc_cutadapt.txt", sep ="/")) %>%
  dplyr::group_by(Sample) %>%
  dplyr::summarize(r_written = sum(r_written), 
            bp_processed = sum(bp_processed),
            quality_trimmed = sum(quality_trimmed),
            bp_written = sum(bp_written))

cont_trim_plot <- trim_stats %>%
  dplyr::select(c("Sample","r_written","bp_processed","quality_trimmed","bp_written")) %>%
  tidyr::pivot_longer(cols = c(3,4)) %>%
  dplyr::mutate(Sample = factor(Sample, levels = trim_stats[order(trim_stats$r_written),]$Sample)
                )

```

```{r cutadapt module}

trim_stats <- read.delim(paste(module_loc,
                               "multiqc",
                               "rnaseq_base_multiqc_data",
                               "multiqc_cutadapt.txt", sep ="/")) %>%
  dplyr::group_by(Sample) %>%
  dplyr::summarize(r_written = sum(r_written), 
            bp_processed = sum(bp_processed),
            quality_trimmed = sum(quality_trimmed),
            bp_written = sum(bp_written))


base_trim_plot <- trim_stats %>%
  dplyr::select(c("Sample","r_written","bp_processed","quality_trimmed","bp_written")) %>%
  tidyr::pivot_longer(cols = c(3,4)) %>%
  dplyr::mutate(Sample = factor(Sample, levels = trim_stats[order(trim_stats$r_written),]$Sample),
                )
```

```{r cutadapt plot}
cont_plot = ggplot(data = cont_trim_plot, aes(x = Sample, y = value, fill = name)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = rev(ggsci::pal_jco("default")(2)),
                    labels = c("BP processed", "BP trimmed")) +
  theme_classic() +
  labs(title = "container" ,
       x = "Samples", 
       y = "Total RNAseq reads", 
       fill = "Read type") +
  theme(axis.text.x = element_blank())

base_plot = ggplot(data = base_trim_plot, aes(x = Sample, y = value, fill = name)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = rev(ggsci::pal_jco("default")(2)),
                    labels = c("BP processed", "BP trimmed")) +
  theme_classic() +
  labs(title = "modules",
       x = "Samples", 
       y = "Total RNAseq reads", 
       fill = "Read type") +
  theme(axis.text.x = element_blank())

ggpubr::ggarrange(base_plot,cont_plot, common.legend = T)

ggsave(filename = "cutadapt_trimming_stats.pdf",
       device = "pdf",
       path = savedir
       )
```




## Alignment

```{r STAR container}

star_stats <- read.delim(paste(container_loc,
                               "multiqc",
                               "rnaseq_container_multiqc_data",
                               "multiqc_star.txt", sep ="/")) %>%
  dplyr::filter(!(grepl("STARpass1", Sample)))

cont_star_perc_stats <- star_stats %>%
  dplyr::select(c("Sample",
                  "uniquely_mapped_percent","multimapped_percent",
                  "unmapped_tooshort_percent","unmapped_other_percent")) %>%
  tidyr::pivot_longer(cols = c(2:5)) %>%
  dplyr::mutate(Sample = factor(Sample, levels = star_stats[order(star_stats$uniquely_mapped_percent),]$Sample),
                name = factor(name,levels = c("unmapped_other_percent","unmapped_tooshort_percent",
                                              "multimapped_percent","uniquely_mapped_percent")))

cont_star_stats <- star_stats %>%
  dplyr::select(c("Sample",
                  "uniquely_mapped", "multimapped", 
                  "unmapped_tooshort", "unmapped_other")) %>%
  tidyr::pivot_longer(cols = c(2:5)) %>%
  dplyr::mutate(Sample = factor(Sample, levels = star_stats[order(star_stats$uniquely_mapped),]$Sample),
                name = factor(name,levels = c("unmapped_other","unmapped_tooshort",
                                              "multimapped","uniquely_mapped")))

```

```{r STAR modules}

star_stats <- read.delim(paste(module_loc,
                               "multiqc",
                               "rnaseq_base_multiqc_data",
                               "multiqc_star.txt", sep ="/")) %>%
  dplyr::filter(!(grepl("STARpass1", Sample)))

base_star_perc_stats <- star_stats %>%
  dplyr::select(c("Sample",
                  "uniquely_mapped_percent","multimapped_percent",
                  "unmapped_tooshort_percent","unmapped_other_percent")) %>%
  tidyr::pivot_longer(cols = c(2:5)) %>%
  dplyr::mutate(Sample = factor(Sample, levels = star_stats[order(star_stats$uniquely_mapped_percent),]$Sample),
                name = factor(name,levels = c("unmapped_other_percent","unmapped_tooshort_percent",
                                              "multimapped_percent","uniquely_mapped_percent")))

base_star_stats <- star_stats %>%
  dplyr::select(c("Sample",
                  "uniquely_mapped", "multimapped", 
                  "unmapped_tooshort", "unmapped_other")) %>%
  tidyr::pivot_longer(cols = c(2:5)) %>%
  dplyr::mutate(Sample = factor(Sample, levels = star_stats[order(star_stats$uniquely_mapped),]$Sample),
                name = factor(name,levels = c("unmapped_other","unmapped_tooshort",
                                              "multimapped","uniquely_mapped")))

```

```{r STAR plot}

cont_plot = ggplot(data = cont_star_stats, aes(x = Sample, y = value, fill = name)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = rev(ggsci::pal_jco("default")(4))) +
  theme_classic() +
  labs(title = "container",
       x = "Samples", 
       y = "Total RNAseq reads", 
       fill = "Read type") +
  theme(axis.text.x = element_blank())

base_plot = ggplot(data = base_star_stats, aes(x = Sample, y = value, fill = name)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = rev(ggsci::pal_jco("default")(4))) +
  theme_classic() +
  labs(title = "module",
       x = "Samples", 
       y = "Total RNAseq reads", 
       fill = "Read type") +
  theme(axis.text.x = element_blank())

ggpubr::ggarrange(base_plot,cont_plot, common.legend = T)

ggsave(filename = "star_alignment_stats.pdf",
       device = "pdf",
       path = savedir
       )

```

```{r STAR perc plot}

cont_perc_plot = ggplot(data = cont_star_perc_stats, aes(x = Sample, y = value, fill = name)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = rev(ggsci::pal_jco("default")(4))) +
  theme_classic() +
  labs(title = "container",
       x = "Samples", 
       y = "Total RNAseq reads", 
       fill = "Read type") +
  theme(axis.text.x = element_blank())

base_perc_plot = ggplot(data = base_star_perc_stats, aes(x = Sample, y = value, fill = name)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = rev(ggsci::pal_jco("default")(4))) +
  theme_classic() +
  labs(title = "module",
       x = "Samples", 
       y = "Total RNAseq reads", 
       fill = "Read type") +
  theme(axis.text.x = element_blank())

ggpubr::ggarrange(base_perc_plot,cont_perc_plot, common.legend = T)

ggsave(filename = "star_percentage_alignment_stats.pdf",
       device = "pdf",
       path = savedir
       )

```


## Stringtie

check stringtie GTF
```{r}
base_gtf <- rtracklayer::import(paste(wd,"analysis","base","stringtie",
                                      "BG96_HM1DCTRL_S13_L001","BG96_HM1DCTRL_S13_L001.gtf",sep="/")) 
container_gtf  <- rtracklayer::import(paste(wd,"analysis","container","stringtie",
                                      "BG96_HM1DCTRL_S13_L001","BG96_HM1DCTRL_S13_L001.gtf",sep="/")) 

length(base_gtf) - length(container_gtf)
```

For some reason, 35 rows in the GTF are different.

```{r }
length(unique(container_gtf$transcript_id))
length(unique(base_gtf$transcript_id))
```

There are 2 transcripts missing in the base GTF

## Stringtie merge

```{r}
base_merged_gffcompare_gtf <- rtracklayer::import(paste(wd,"analysis",
                                                        "base","gffcompare","base_merged","base_merged_gffcompare.annotated.gtf", sep = "/"))

container_merged_gffcompare_gtf <- rtracklayer::import(paste(wd,"analysis",
                                                        "container","gffcompare","container_pancreas","container_pancreas_gffcompare.annotated.gtf", sep = "/"))

length(base_merged_gffcompare_gtf) - length(container_merged_gffcompare_gtf)
```
Now we suddenly have a difference of 1400 lines in the Granges, where the base
has more far more lines. This is the biggest difference so far

```{r}
length(unique(base_merged_gffcompare_gtf$transcript_id))
length(unique(container_merged_gffcompare_gtf$transcript_id))
```

Now, we miss 76 transcripts that are not present in the container pipeline 
compared to the base pipeline

```{r}
base_novel_df <- as.data.frame(base_merged_gffcompare_gtf) %>%
  dplyr::filter(!(class_code == "="))
  
container_novel_df <- as.data.frame(container_merged_gffcompare_gtf) %>%
  dplyr::filter(!(class_code == "="))

table(base_novel_df$class_code)
table(container_novel_df$class_code)
```

## Filtered GTF

```{r}
base_filter_gtf <- rtracklayer::import(paste(wd,"analysis","base","customannotation",
                                             "base_merged_novel_filtered.gtf", sep = "/")
)

container_filter_gtf <- rtracklayer::import(paste(wd,"analysis","container","customannotation",
                                             "container_pancreas_novel_filtered.gtf", sep = "/")
)

length(base_filter_gtf) - length(container_filter_gtf)
```

Now, there is a different of 56 more lines in the container version (in 3 million, that is not a lot)

```{r}
length(unique(base_filter_gtf$transcript_id))

length(unique(container_filter_gtf$transcript_id))
```

the container version has 2 more TXs

```{r missing TXs}
base_diff <- setdiff(base_filter_gtf, container_filter_gtf)
based_diff_df <- as.data.frame(base_diff)
container_diff <- setdiff(container_filter_gtf,base_filter_gtf)
```

In total, there are 5 ranges extra ranges in container DF and none in base diff. 
This seems small enough to disregard

```{r}
base_novel_df <- as.data.frame(base_filter_gtf) %>%
  dplyr::filter(!(class_code == "="))
base_novel_df <- as.data.frame(table(base_novel_df$class_code)) %>%
  dplyr::mutate(type = "base")

container_novel_df <- as.data.frame(container_filter_gtf) %>%
  dplyr::filter(!(class_code == "="))
container_novel_df <- as.data.frame(table(container_novel_df$class_code)) %>%
  dplyr::mutate(type = "container")

plot_df <- rbind(container_novel_df, base_novel_df)
colnames(plot_df) <- c("class_code","count","type")

base_df <- as.data.frame(table(base_novel_df$class_code)
table(container_novel_df$class_code)
```

```{r plot difference}

ggplot(plot_df, aes(x = class_code, y = count, fill = type,
                    group = type, label = count,
                    after_scale(alpha(fill, 0.4)))) +
  geom_bar(position = "dodge",stat = "identity") +
  geom_text(stat='identity', position = position_dodge(width = .9), vjust = -.3, size = 3.5) +
  scale_fill_manual(values = viridis::viridis(2, begin = .2, end = .6)) +
  theme_classic() +
  scale_y_log10() +
  theme(axis.text.y=element_blank(),axis.ticks=element_blank(),
          axis.title.y=element_blank())

ggsave(filename = paste(wd,"results","comparison_analysis","merged_annotated_gtf_comparison.pdf", sep ="/"))

```

## GFFcompare GTF

```{r }
compare_gtf <- rtracklayer::import(paste(wd,"analysis","gffcompare","pipeline_comparison","container_vs_base_comparison_gffcompare.annotated.gtf",sep="/"))

compare_df <- as.data.frame(compare_gtf) %>%
  dplyr::filter(type == "transcript" & class_code %in% c("j","k"))
table(compare_df$class_code)

```

Only 7 transcripts have a different exon structure according to GFFcompare

**After manual inspection, they look better in the container version than in the**
**container version.**